FontSizePicker: Use number values when the initial value is a number

Mamaduka opened the pull request:
## Description
Updates `onChange` logic in FontSizePicker component. When the component detects that value or first object in `fonSizes` doesn't use units, it will pass the number value to the onChange callback.

Fixes #33651.
Fixes #33653.

## How has this been tested?
Tested with plugin provided in original issue report - https://github.com/NasKadir123/textBlock

## Types of changes
Bugfix 

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

Mamaduka committed:
withFontSizes: Parse units into a numbers

olafleur commented:
Thanks! :)

youknowriad commented:
Is it possible to fix the issue in `FontSizePicker` instead?

youknowriad commented:
Asking because this will fix `withFontSizes` but `FontSizePicker` direct usage still remains broken right?

Mamaduka commented:
> Asking because this will fix withFontSizes but FontSizePicker direct usage still remains broken right?

The `FontSizePicker` now will return units and blocks can use those directly in attributes - #31314.

If we have to parse units into numbers there as well, then using `UnitControl` doesn't really make sense.

youknowriad commented:
I was thinking something like: if the passed (previous) value passed to `FontSizePicker` is a number and the selected unit is "px", call onChange with just the number (and not the full string).

That said, I still prefer this PR because the proposal is a hack but the hack might be necessary depending on the scale of the breaking change.

youknowriad reviewed (approved):
Maybe this is good enough as both reported issues are specifically about withFontSizes

Mamaduka commented:
So far, I've only seen reports about `withFontSizes`. It might be better to add a proposed hack if we see reports for the `FontSizePicker` component.

Side note: The problem can also be fixed by changing attribute type from "number" to "string."

youknowriad commented:
One question though about the fix here. It seems that the proposed fix assumes that only `number` attributes are supported with `withFontSize`. Won't that break if we use that with the default `FontSizePicker`. Meaning someone could select "2em" but it would get applied as "2px" ?

Mamaduka commented:
The `withFontSize` was only using numbers before, so users/devs expect numbers in their attributes.

But yeah, sadly, unit control values will have no effect.

youknowriad commented:
What was the behavior in WP 5.7?

Mamaduka commented:
The `FontSizePicker` was using `NumberControl`, so onChange was returning numbers.

What do you think about this idea?
* Use `parseCSSUnitValue` to parse values.
* If the unit is `px`, return parsed number, otherwise return the whole unit.

Mamaduka commented:
Currently, I can think of the one issue with my proposed solution: if the user changes unit and attribute type is a number, it will trigger block validation.

youknowriad commented:
> The FontSizePicker was using NumberControl, so onChange was returning numbers.

So we introduced units only in 5.7? In my mind, we did introduce that before but maybe I'm wrong?

Mamaduka commented:
I think we introduced the units before, but we update NumberControl to UnitControl in FontSizePicker in 5.8. This is the PR - #31314.

youknowriad commented:
Ok, so it seems the best approach would be to revert the FontSizePicker to `pixel` only (to avoid any breaking changes) and have a way to trigger multi-units by padding a `units` array or something like that (or potentially a boolean) which we should enable for core blocks all hooks. Is something like that possible?

Also curious about @aristath's opinion here.

aristath commented:
Making the `FontSizePicker` use `UnitControl` instead of `NumberControl` was a significant improvement... Reverting that would be a big regression for designers.
What if, instead of reverting an improvement, we introduced a package/method to automatically convert all units to pixel values when needed? It should be possible to convert `2em` to `32px` and get `32` as a number...

Mamaduka commented:
It looks like we already have `hasUnits` check for values:

https://github.com/WordPress/gutenberg/blob/ed8a7ad7ab127d59a1407941c2ec1e639bb440e6/packages/components/src/font-size-picker/index.js#L66-L68

I think we can use that check and do following:
```js
onChange( hasUnits ? nextSize : parseInt( nextSize, 10 ) );
```

This way, we'll consider the block's current attribute value and pre-defined sizes, allowing us to use units or numbers.

I will update PR shortly to test this case.

youknowriad commented:
@aristath I'm not suggesting to revert but to make it optional (and enabled for all Core things) because switching the value from number to attribute is breaking all plugins that rely on the "value" of FontSizePicker being a number.

Mamaduka committed:
Revert "withFontSizes: Parse units into a numbers"

This reverts commit 5ea6028522d5c353d4defd84d22b713f31bd991a.

Mamaduka committed:
FontSizePicker: Don't use units if the value is a number

Mamaduka commented:
@youknowriad, @aristath

Removed original fix and updated `onChange` logic in FontSizePicker component. When the component detects that `value` or first object in `fonSizes` doesn't use units, it will pass the number to the onChange callback.

It should fix the BC issue and allow developers to use units with FontSizePicker.

youknowriad commented:
@Mamaduka maybe I'm asking too much but I feel like we should have an e2e test with the provided plugin to avoid future regressions. WDYT?

Mamaduka commented:
I think this case is more suited for unit tests - `it should return number if value|sizes are numbers`.

Or do you want to test that it doesn't trigger block validation and attributes are correctly defined?.

youknowriad commented:
Whatever works best for you :)

Mamaduka committed:
Add unit tests

Mamaduka commented:
@aristath, @youknowriad

Added unit tests.

youknowriad commented on packages/components/src/font-size-picker/test/index.js:
```diff
@@ -0,0 +1,115 @@
+
+describe( 'FontSizePicker', () => {
+	describe( 'onChange values', () => {
+		it( 'should not use units when the initial value is a number', () => {
```
Should we check that the units input is not render or do you think it should be rendered in this case?

Mamaduka replied:
Do you mean units "switcher"? It's is still rendered but has no effect since we only use number values.

I think toggling between `NumberControl` and `UnitControl` based on has units value will introduce more complexity to the component.

![CleanShot 2021-07-30 at 13 37 06](https://user-images.githubusercontent.com/240569/127634040-cc46b175-c3d2-425a-aec1-d59f62b320af.png)

youknowriad replied:
> I think toggling between NumberControl and UnitControl based on has units value will introduce more complexity to the component.

Can't we just pass an empty array of units or something like that in that case? It feels weird for a user to select "em" but have pixels applied.

Mamaduka replied:
Done. It will now only display the "PX" unit label when there are no units.

Mamaduka committed:
Disable units when we have number values

youknowriad commented:
This is great for me, thanks for all the iterations.

Mamaduka commented:
Not a problem. I'm just happy we were able to find the right way to fix this issue.
